### PR TITLE
feat!: update data type in ResourceProperties

### LINKED ISSUE #770: [EdgeX 3.0] Device Profile model improvement
# 🚀 Feature Request

### Relevant Package [**REQUIRED**]

 This feature request is for core-metadata

### Description [**REQUIRED**]
This is issue is opened for discussing the Device Profile model improvment in v3.0.

### Describe the solution you'd like
The first suggestion is that Value Properties base, offset, scale etc become numerics rather than strings

### PR DESCRIPTION
**Pull request by hahattan**
BREAKING CHANGE: update mask,shift,base,scale,offset to numeric data type

also rename 'Others' to 'Optional' for the consistency with message broker settings.

closes #770

If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-core-contracts/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [ ] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

### CODE REVIEW

File: dtos/deviceprofile_test.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ var testDeviceProfile = models.DeviceProfile{
 		Properties: models.ResourceProperties{
 			ValueType: common.ValueTypeInt16,
 			ReadWrite: common.ReadWrite_RW,
-			Others:    testTags,
+			Optional:  testTags,
```
Comment (lenny-intel) [inline dtos/deviceprofile_test.go:40]: Why using `testTags`? Optional is not for tags, so please use something more relevant to what we expect it to be used for.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Review by lenny-intel — APPROVED** [review]
LGTM

**Review by iain-anderson — CHANGES_REQUESTED** [review]
Can we use json.Number for scale, offset and base - so that they can be integer or float?

**Comment by hahattan** [comment]
> Can we use json.Number for scale, offset and base - so that they can be integer or float?

`json.Number` is alias of `string` which provides useful methods to transform value to numeric data type.
But there still will be error in the same place if user provides illegal value:
https://github.com/edgexfoundry/device-sdk-go/blob/b46b419fd49731d0deb4deed8e6127ef6a42900c/internal/transformer/transformresult.go#L92-L93

Using `float64` directly can avoid the error during device profile creation but it's a breaking change.

Another way to improve this is to use `json.Number` for  these properties, but leverage validator to check the type.
For example https://pkg.go.dev/github.com/go-playground/validator/v10#hdr-Numeric for `base`, `scale` and `offset`.
But we might need to implement custom tag for `mask` and `shift`

@lenny-intel @cloudxxx8 @iain-anderson thought ?

**Comment by cloudxxx8** [comment]
@iain-anderson is it ok?

**Comment by iain-anderson** [comment]
OK with this as is

**Pull request merged**